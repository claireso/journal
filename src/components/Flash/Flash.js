import React, { useEffect, useRef } from 'react'
import PropTypes from 'prop-types'

import { IconClose } from '@components/Icons'

import * as S from './Flash.styles'

const Flash = ({ status, children, onClose, index, withBorder, ...props }) => {
  const dom = useRef(null)

  useEffect(() => {
    if (dom.current?.animate) {
      dom.current.animate([{ opacity: 0 }, { opacity: 1 }], {
        duration: 1000,
        delay: 50,
        easing: 'cubic-bezier(.17,.67,.21,.97)',
        fill: 'forwards'
      })
    }
  }, [])

  return (
    <S.Wrapper ref={dom} status={status} separator={withBorder} {...props}>
      {children}
      {onClose && (
        <S.ButtonClose
          onClick={(event) => {
            event.preventDefault()
            onClose(index)
          }}
        >
          <IconClose />
        </S.ButtonClose>
      )}
    </S.Wrapper>
  )
}

Flash.propTypes = {
  children: PropTypes.node.isRequired,
  index: PropTypes.number,
  status: PropTypes.string,
  onClose: PropTypes.func,
  withBorder: PropTypes.bool
}

Flash.defaultProps = {
  index: 0,
  status: 'default',
  onClose: () => {},
  withBorder: false
}

export default Flash
