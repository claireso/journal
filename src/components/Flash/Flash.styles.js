import { styled } from '@theme'

import { Button } from '../Buttons'

export const Wrapper = styled('div', {
  fontSize: '$4',
  mb: '$5',
  p: '$5',
  position: 'relative',
  textAlign: 'center',
  opacity: 0,

  variants: {
    status: {
      default: {
        backgroundColor: '$primary100',
        color: '$white',
        borderColor: '$primary200'
      },
      error: {
        background: '$error100',
        color: '$white',
        borderColor: '$error200'
      },
      success: {
        background: '$success100',
        color: '$white',
        borderColor: '$success200'
      },
      info: {
        backgroundColor: '$tertiary100',
        color: '$gray600',
        borderColor: '$tertiary200'
      }
    },
    separator: {
      true: {
        borderTopStyle: 'solid',
        borderTopWidth: '3px'
      }
    }
  }
})

export const ButtonClose = styled(Button, {
  p: '0 $5',
  alignItems: 'center',
  display: 'flex',
  position: 'absolute',
  right: 0,
  top: 0,
  bottom: 0,
  '> svg': {
    fill: 'currentColor',
    m: 0
  }
})
