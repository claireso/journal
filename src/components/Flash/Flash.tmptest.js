//#TODO restore test

// import { render, fireEvent } from '@testing-library/react'

// import Flash from './index'

// describe('<Flash />', () => {
//   test('should render default flash message', () => {
//     const { container } = render(<Flash>Flash message</Flash>)

//     expect(container).toMatchSnapshot()
//   })

//   test('should render success flash message', () => {
//     const { container } = render(<Flash status="success">Flash message</Flash>)

//     expect(container).toMatchSnapshot()
//   })

//   test('should render error flash message', () => {
//     const { container } = render(<Flash status="error">Flash message</Flash>)

//     expect(container).toMatchSnapshot()
//   })

//   test('should close flash message', () => {
//     const spyOnClose = jest.fn()

//     const { container } = render(<Flash onClose={spyOnClose}>Flash message</Flash>)

//     expect(container).toMatchSnapshot()

//     const closeButton = container.querySelector('button')

//     fireEvent.click(closeButton)

//     expect(spyOnClose).toHaveBeenCalledTimes(1)
//   })
// })
