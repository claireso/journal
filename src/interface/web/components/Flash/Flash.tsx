'use client'

import React, { useEffect, useRef } from 'react'

import Icon from '../Icons'

import type { FlashVariants } from './styles.css'
import * as cls from './styles.css'
import joinCls from '@utils/joinCls'

type FlashProps = {
  children: React.ReactNode
  index?: number
  withBorder?: boolean
  onClose?: ((index: number) => void) | null
  className?: string
} & FlashVariants

const Flash = ({
  status = 'default',
  children,
  onClose = null,
  index = 0,
  withBorder = false,
  className
}: FlashProps) => {
  const dom = useRef<HTMLDivElement>(null)

  useEffect(() => {
    if (dom.current?.animate) {
      dom.current.animate([{ opacity: 0 }, { opacity: 1 }], {
        duration: 1000,
        delay: 50,
        easing: 'cubic-bezier(.17,.67,.21,.97)',
        fill: 'forwards'
      })
    }
  }, [])

  return (
    <div ref={dom} className={joinCls([cls.wrapper({ status, separator: withBorder }), className])}>
      {children}
      {onClose && (
        <button
          className={cls.buttonClose}
          onClick={(event) => {
            event.preventDefault()
            onClose(index)
          }}
        >
          <Icon name="close" />
        </button>
      )}
    </div>
  )
}

export default Flash
