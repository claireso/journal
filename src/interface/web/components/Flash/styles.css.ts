import { recipe, RecipeVariants } from '@vanilla-extract/recipes'
import { cmpStyles, p, px, tokens } from '@web/theme'

export const wrapper = recipe({
  base: cmpStyles({
    position: 'relative',
    textAlign: 'center',
    opacity: 0,
    fontSize: tokens.fonts.size.sm,
    ...p('size-5')
  }),
  variants: {
    status: {
      default: cmpStyles({
        color: tokens.colors.primary.base,
        background: tokens.colors.primary.extralight,
        borderColor: tokens.colors.primary.light
      }),
      error: cmpStyles({
        color: tokens.colors.danger.base,
        background: tokens.colors.danger.extralight,
        borderColor: tokens.colors.danger.light
      }),
      success: cmpStyles({
        color: tokens.colors.success.base,
        background: tokens.colors.success.extralight,
        borderColor: tokens.colors.success.light
      }),
      info: cmpStyles({
        color: tokens.colors['text-base'],
        background: tokens.colors.info.extralight,
        borderColor: tokens.colors.info.light
      })
    },
    separator: {
      true: cmpStyles({
        borderTopStyle: 'solid',
        borderTopWidth: '2px'
      })
    }
  },
  defaultVariants: {
    status: 'default',
    separator: false
  }
})

export type FlashVariants = RecipeVariants<typeof wrapper>

export const buttonClose = cmpStyles({
  display: 'flex',
  alignItems: 'center',
  position: 'absolute',
  right: 0,
  top: 0,
  bottom: 0,
  ...px('size-5')
})
