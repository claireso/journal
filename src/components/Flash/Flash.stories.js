import React, { Fragment } from 'react'
import { action } from '@storybook/addon-actions'

import Flash from './index'

export default {
  title: 'Flash'
}

export const Basic = () => (
  <Fragment>
    <div>
      <Flash>Flash message</Flash>
    </div>
    <div>
      <Flash onClose={action('close message')}>Closable flash message</Flash>
    </div>
  </Fragment>
)

export const Success = () => (
  <Fragment>
    <div>
      <Flash status="success">Flash message</Flash>
    </div>
    <div>
      <Flash status="success" onClose={action('close message')}>
        Closable flash message
      </Flash>
    </div>
  </Fragment>
)

export const Error = () => (
  <Fragment>
    <div>
      <Flash status="error">Flash message</Flash>
    </div>
    <div>
      <Flash status="error" onClose={action('close message')}>
        Closable flash message
      </Flash>
    </div>
  </Fragment>
)
