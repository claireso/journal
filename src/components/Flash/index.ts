import Flash from './Flash'

export default Flash
