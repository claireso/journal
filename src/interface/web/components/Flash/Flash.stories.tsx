import React, { Fragment } from 'react'
import { action } from 'storybook/actions'

import type { Meta, StoryObj } from '@storybook/react-webpack5'

import Flash from './Flash'

const meta: Meta<typeof Flash> = {
  title: 'Components/Flash',
  component: Flash,
  args: {
    onClose: action('close message')
  },
  render({ status, onClose }) {
    return (
      <Fragment>
        <div>
          <Flash status={status}>Flash message</Flash>
        </div>
        <div>
          <Flash status={status} onClose={onClose}>
            Closable flash message
          </Flash>
        </div>
      </Fragment>
    )
  }
}

export default meta
type Story = StoryObj<typeof Flash>

export const Basic: Story = {
  args: {
    status: 'default'
  }
}

export const Success: Story = {
  args: {
    status: 'success'
  }
}

export const Error: Story = {
  args: {
    status: 'error'
  }
}

export const Info: Story = {
  args: {
    status: 'info'
  }
}
