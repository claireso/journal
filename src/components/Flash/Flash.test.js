import { render, screen, fireEvent } from '@testing-library/react'

import Flash from './index'

describe('<Flash />', () => {
  it('should render default message', () => {
    const { asFragment } = render(<Flash>Flash message</Flash>)

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render success message', () => {
    const { asFragment } = render(<Flash status="success">Flash message</Flash>)

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render error message', () => {
    const { asFragment } = render(<Flash status="error">Flash message</Flash>)

    expect(asFragment()).toMatchSnapshot()
  })

  it('should close message', () => {
    const onClose = jest.fn()

    const { asFragment } = render(<Flash onClose={onClose}>Flash message</Flash>)

    expect(asFragment()).toMatchSnapshot()

    fireEvent.click(screen.getByRole('button'))

    expect(onClose).toHaveBeenCalledTimes(1)
  })
})
