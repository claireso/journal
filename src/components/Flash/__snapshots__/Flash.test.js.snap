// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Flash /> should close message 1`] = `
<DocumentFragment>
  <div
    class="c-fyLyTf c-fyLyTf-kioelQ-status-default"
  >
    Flash message
    <button
      class="c-ggYOOi c-eUyIYb c-hrWJbv"
    >
      <svg
        height="16"
        viewBox="0 0 22 28"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20.281 20.656c0 0.391-0.156 0.781-0.438 1.062l-2.125 2.125c-0.281 0.281-0.672 0.438-1.062 0.438s-0.781-0.156-1.062-0.438l-4.594-4.594-4.594 4.594c-0.281 0.281-0.672 0.438-1.062 0.438s-0.781-0.156-1.062-0.438l-2.125-2.125c-0.281-0.281-0.438-0.672-0.438-1.062s0.156-0.781 0.438-1.062l4.594-4.594-4.594-4.594c-0.281-0.281-0.438-0.672-0.438-1.062s0.156-0.781 0.438-1.062l2.125-2.125c0.281-0.281 0.672-0.438 1.062-0.438s0.781 0.156 1.062 0.438l4.594 4.594 4.594-4.594c0.281-0.281 0.672-0.438 1.062-0.438s0.781 0.156 1.062 0.438l2.125 2.125c0.281 0.281 0.438 0.672 0.438 1.062s-0.156 0.781-0.438 1.062l-4.594 4.594 4.594 4.594c0.281 0.281 0.438 0.672 0.438 1.062z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`<Flash /> should render default message 1`] = `
<DocumentFragment>
  <div
    class="c-fyLyTf c-fyLyTf-kioelQ-status-default"
  >
    Flash message
  </div>
</DocumentFragment>
`;

exports[`<Flash /> should render error message 1`] = `
<DocumentFragment>
  <div
    class="c-fyLyTf c-fyLyTf-epOfjx-status-error"
  >
    Flash message
  </div>
</DocumentFragment>
`;

exports[`<Flash /> should render success message 1`] = `
<DocumentFragment>
  <div
    class="c-fyLyTf c-fyLyTf-eIkjtV-status-success"
  >
    Flash message
  </div>
</DocumentFragment>
`;
